Add workflow documentation (#40)
exercism/development-environment: docs/creating_new_react_component.md

# Creating a new React Component for v3's website

> 

Author: ```suggestion

For our purposes, the component will be named `<ComponentName>` with different case combinations used in their specific context. When working through this, replace `<ComponentName>` with your component's name.

```

## Create the component

- Each component should be in its own directory

> - Should be placed in `/app/javascript/components/<componentname>`

Author: ```suggestion
- Should be placed in `/app/javascript/components/<component-name>`
```

## Create the tests

> - Place tests in `/test/javascript/components/<component-name>/ComponentName.test.tsx`

Reviewer: I suggest prefixing/suffixing `ComponentName` with `$`, e.g. `$ComponentName$`, to make it clear that it's a placeholder. And doing the same for `$component_name$` further down. A different placeholder syntax would be fine too as long as it conveys the message 🙂

Reviewer 2: I am usually a fan of `<` and `>` as placeholders, but feel free to use whatever placeholder you prefer.

Reviewer: Yeah, `<>` are better than what I suggested!

> - The tests must use the [JEST](https://jestjs.io/) test framework

Reviewer 2: ```suggestion
- The tests must use the [Jest](https://jestjs.io/) test framework
```

Author: I wasn't sure, and the website has it all in caps. :P thanks!

- [Testing Library](https://testing-library.com/) is useful for testing your react component

> - [Pluralsight has a crashcourse](https://www.pluralsight.com/guides/how-to-test-react-components-in-typescript) in testing React component

Reviewer 2: ```suggestion
- [Pluralsight has a crash course](https://www.pluralsight.com/guides/how-to-test-react-components-in-typescript) in testing React components
```

- See [common workflows](./common_workflows.md) to see how to run the jest tests

## Add it to the application

- Open `/app/javascript/packs/application.tsx`
- The `initReact` function takes an object where each entry represents a react component
    - The key is a string representing your component
    - The value is a function
        - Arguments
            - `data`, represents the data required by your component to be initialized
        - Return
            - Your react component, with the `data` assigned to it
    - assume the data has already been massaged to a desired format

## Add the a test route

- Open `/config/routes.rb`
- In `namespace :test` > `namespace :components` add your component to the proper section adding a `:show` resource

> `resource  :component_name, only: \[:show\], controller:  'component_name'`

Author: ```suggestion
`resource  :<component_name>, only: \[:show\], controller:  '<component_name>'`
```

## Create the test controller

> - Create a file `/app/controllers/test/components/component_name_controller.rb`

Author: ```suggestion
- Create a file `/app/controllers/test/components/<component_name>_controller.rb`
```

- Create a class that inherits from the `ApplicationController`

> - This allows you to navigate in the browser to: `http://localhost:3020/test/components/component_name` to view the component rendered

Author: ```suggestion
- This allows you to navigate in the browser to: `http://localhost:3020/test/components/<component_name>` to view the component rendered
```

```ruby

> class  Test::Components::ComponentName < ApplicationController  \# rubocop:disable Layout/LineLength

Author: ```suggestion
class  Test::Components::<ComponentName> < ApplicationController  \# rubocop:disable Layout/LineLength
```

    def show
        @data = {
            # Whatever mocked data should be passed to your component to view it
        }
    end
end
```

## Create the view

React components at this time are used to supplement other views, and likely do not have a view for you to create at this time

## Create the view helper test

> - Create a file `/test/helpers/view_components/component_name_test.rb`

Author: ```suggestion
- Create a file `/test/helpers/view_components/<component_name>_test.rb`
```

- This tests that the output from the Route renders matching HTML generated by `application.tsx`

```ruby
require_relative "../view_component_test_case"

> class ComponentNameTest < ViewComponentTestCase

Author: ```suggestion
class <ComponentName>Test < ViewComponentTestCase
```

    test "component with empty concepts map rendered correctly" do
      data = {
          # Whatever data you want to test
      }

>       component = ViewComponents::ComponentName.new(data).to_s

Author: ```suggestion
      component = ViewComponents::<ComponentName>.new(data).to_s
```

>       assert_component component, "concepts-map", data

Author: ```suggestion
      assert_component component, "<component-name>", data
```

    end
end
```

## Create the system test (for Capybara)

> - Create a file `/test/system/component/component_name_test.rb`

Author: ```suggestion
- Create a file `/test/system/component/<component_name>_test.rb`
```

```ruby
require "application_system_test_case"

module Components

>     class ComponentNameTest < ApplicationSystemTestCase

Author: ```suggestion
    class <ComponentName>Test < ApplicationSystemTestCase
```
